File: Hackerearth/February Circuits/biject_inside.cpp
/*
                C++
                encoding:UTF-8
                Modified: <20/Feb/2019 05:43:54 PM>

                ✪ H4WK3yE乡
                Mohd. Farhan Tahir
                Indian Institute Of Information Technology (IIIT),Gwalior
*/

#include <bits/stdc++.h>
using namespace std;

#define ll long long
#define mod 1000000007
#define sum(v) accumulate(v.begin(), v.end(), 0);
#define reset(a, b) memset(a, b, sizeof(a));
#define pb push_back

ll powermod(ll _a, ll _b, ll _m) {
  ll _r = 1;
  while (_b) {
    if (_b % 2 == 1)
      _r = (_r * _a) % _m;
    _b /= 2;
    _a = (_a * _a) % _m;
  }
  return _r;
}

void multiply(ll F[][2], ll M[][2]) {
  ll x = (((F[0][0] % mod * M[0][0] % mod) % mod) +
          ((F[0][1] % mod * M[1][0] % mod) % mod)) %
         mod;
  ll y = (((F[0][0] % mod * M[0][1] % mod) % mod) +
          ((F[0][1] % mod * M[1][1] % mod) % mod)) %
         mod;
  ll z = (((F[1][0] % mod * M[0][0] % mod) % mod) +
          ((F[1][1] % mod * M[1][0] % mod) % mod)) %
         mod;
  ll w = (((F[1][0] % mod * M[0][1] % mod) % mod) +
          ((F[1][1] % mod * M[1][1] % mod) % mod)) %
         mod;
  F[0][0] = x, F[0][1] = y, F[1][0] = z, F[1][1] = w;
}

void power(ll matrix[2][2], ll n) {
  if (n == 0 || n == 1)
    return;
  ll res[2][2] = {{1, 1}, {1, 0}};
  power(matrix, n / 2);
  multiply(matrix, matrix);
  if (n & 1)
    multiply(matrix, res);
}

ll fib(ll n) {
  ll matrix[2][2] = {{1, 1}, {1, 0}};
  power(matrix, n - 1);
  return matrix[0][0];
}

int main() {
  // ios_base::sync_with_stdio(false), cin.tie(0), cout.tie(0);
  int tc;
  cin >> tc;
  ll n;
  while (tc--) {
    cin >> n;
    ll v = powermod(2, n, mod);
    ll u = (v - fib(n + 1) % mod + mod) % mod;
    cout << u << " " << v << "\n";
    ll ans = (((u % mod) * powermod(v, mod - 2, mod)) % mod);
    cout << ans << "\n";
  }
  return 0;
}

File: Hackerearth/February Circuits/biject_inside.py
#	!/usr/bin/env python3
#	coding: UTF-8
#	Modified: <20/Feb/2019 05:54:26 PM>


#	✪ H4WK3yE乡
#	Mohd. Farhan Tahir
#	Indian Institute Of Information Technology (IIIT),Gwalior

#	Question Link
#
#

# ///==========Libraries, Constants and Functions=============///


import sys
sys.setrecursionlimit(10**6)
inf = float("inf")
mod = 1000000007


def get_array(): return list(map(int, sys.stdin.readline().split()))


def get_ints(): return map(int, sys.stdin.readline().split())


def input(): return sys.stdin.readline()

# ///==========MAIN=============///

'''
def multiply(F, M):
    x = (((F[0][0]*M[0][0]) % mod) + ((F[0][1]*M[1][0]) % mod)) % mod
    y = (((F[0][0]*M[0][1]) % mod) + ((F[0][1]*M[1][1]) % mod)) % mod
    z = (((F[1][0]*M[0][0]) % mod) + ((F[1][1]*M[1][0]) % mod)) % mod
    w = (((F[1][0]*M[0][1]) % mod) + ((F[1][1]*M[1][1]) % mod)) % mod

    F[0][0] = x
    F[0][1] = y
    F[1][0] = z
    F[1][1] = w


def power(matrix, n):
    if n == 0 or n == 1:
        return
    res = [[1, 1], [1, 0]]
    power(matrix, n//2)
    multiply(matrix, matrix)
    if n & 1:
        multiply(matrix, res)


def fib(n):
    matrix = [[1, 1], [1, 0]]
    if n == 0:
        return 0
    power(matrix, n-1)
    return matrix[0][0]
'''

import math

sqrt_five = math.sqrt(5)
Phi = (1 + sqrt_five)/2


def fib(num):
    return int(((pow(Phi, num, mod))-(pow(-Phi, -num, mod)))/sqrt_five)


def main():
    '''from itertools import permutations, combinations
    n = int(input())
    arr = [i+1 for i in range(n)]
    p = []
    for i in permutations(arr):
        p.append(i)
    # print(p)
    s = []
    for i in range(n+1):
        x = combinations(arr, i)
        for j in x:
            s.append(j)
    # print(s)
    ans = 0
    for k in s:
        for i in p:
            for j in range(n):
                if (j+1 in k and i[j] in k):
                    ans += 1
                    break
    from math import gcd, factorial
    d = factorial(n)
    while gcd(d, ans) != 1:
        u = gcd(d, ans)
        d //= u
        ans //= u
    print(ans, d)

    numerator== 2**n-fib(n)'''
    for tc in range(int(input())):
        n = int(input())
        y = pow(2, n, mod)
        x = y-fib(n+1)
        # print(fib(n+1))
        print(x, y)
        print(((x % mod) * pow(y, mod - 2, mod)) % mod)


if __name__ == "__main__":
    main()
